rust/xaynet-core/src/mask/masking.rs

```diff
@@ -363,23 +355,20 @@ impl Masker {
     /// [`unmask()`]: struct.Aggregation.html#method.unmask
     pub fn mask(self, scalar: f64, model: Model) -> (MaskSeed, MaskObject) {
         let (random_int, mut random_ints) = self.random_ints();
-        let Self {
-            config_model,
-            config_scalar,
-            seed,
-        } = self;
+        let Self { config, seed } = self;
+        let (config_n, config_1) = (config.vect, config.unit);
```

Comment:
Changed to destructure as suggested (although I'm not entirely sure that that gives it move semantics either?).

Reply:
i looked further into that and it seems to be indeed the case that it's copied instead of moved, at least that's what i get from the following example compiled into its MIR representation.

the example has a simplistic setup with a struct `C` holding a `Clone` value of `A` and a `Copy` value of `B`:

```rust
#[derive(Clone)]
struct A {
    a: usize,
}

#[derive(Copy, Clone)]
struct B {
    b: usize,
}

#[derive(Clone)]
struct C {
    a: A,
    b: B,
}

fn main() {
    let a = A { a: rand::random() };
    let b = B { b: rand::random() };
    let c = C { a, b };
    
    let C { a: c1, b: c2 } = c;
}
```

the call to random is needed to prevent some premature optimization, otherwise it will just propagate some `const`. in the following MIR output you can see that the value `c1` from the destructuring gets cloned (see `_8 = move (_5.0: A);` in block `bb2`), whereas the value `c2` gets copied (see `_9 = (_5.1: B);` in block `bb2`):

```
fn main() -> () {
    let mut _0: ();                      // return place in scope 0 at src/main.rs:17:11: 17:11
    let _1: A;                           // in scope 0 at src/main.rs:18:9: 18:10
    let mut _2: usize;                   // in scope 0 at src/main.rs:18:20: 18:34
    let mut _4: usize;                   // in scope 0 at src/main.rs:19:20: 19:34
    let mut _6: A;                       // in scope 0 at src/main.rs:20:17: 20:18
    let mut _7: B;                       // in scope 0 at src/main.rs:20:20: 20:21
    scope 1 {
        debug a => _1;                   // in scope 1 at src/main.rs:18:9: 18:10
        let _3: B;                       // in scope 1 at src/main.rs:19:9: 19:10
        scope 2 {
            debug b => _3;               // in scope 2 at src/main.rs:19:9: 19:10
            let _5: C;                   // in scope 2 at src/main.rs:20:9: 20:10
            scope 3 {
                debug c => _5;           // in scope 3 at src/main.rs:20:9: 20:10
                let _8: A;               // in scope 3 at src/main.rs:22:16: 22:18
                let _9: B;               // in scope 3 at src/main.rs:22:23: 22:25
                scope 4 {
                    debug c1 => _8;      // in scope 4 at src/main.rs:22:16: 22:18
                    debug c2 => _9;      // in scope 4 at src/main.rs:22:23: 22:25
                }
            }
        }
    }

    bb0: {
        StorageLive(_1);                 // scope 0 at src/main.rs:18:9: 18:10
        StorageLive(_2);                 // scope 0 at src/main.rs:18:20: 18:34
        _2 = rand::random::<usize>() -> bb1; // scope 0 at src/main.rs:18:20: 18:34
                                         // mir::Constant
                                         // + span: src/main.rs:18:20: 18:32
                                         // + literal: Const { ty: fn() -> usize {rand::random::<usize>}, val: Value(Scalar(<ZST>)) }
    }

    bb1: {
        (_1.0: usize) = move _2;         // scope 0 at src/main.rs:18:13: 18:36
        StorageDead(_2);                 // scope 0 at src/main.rs:18:35: 18:36
        StorageLive(_3);                 // scope 1 at src/main.rs:19:9: 19:10
        StorageLive(_4);                 // scope 1 at src/main.rs:19:20: 19:34
        _4 = rand::random::<usize>() -> bb2; // scope 1 at src/main.rs:19:20: 19:34
                                         // mir::Constant
                                         // + span: src/main.rs:19:20: 19:32
                                         // + literal: Const { ty: fn() -> usize {rand::random::<usize>}, val: Value(Scalar(<ZST>)) }
    }

    bb2: {
        (_3.0: usize) = move _4;         // scope 1 at src/main.rs:19:13: 19:36
        StorageDead(_4);                 // scope 1 at src/main.rs:19:35: 19:36
        StorageLive(_5);                 // scope 2 at src/main.rs:20:9: 20:10
        StorageLive(_6);                 // scope 2 at src/main.rs:20:17: 20:18
        _6 = move _1;                    // scope 2 at src/main.rs:20:17: 20:18
        StorageLive(_7);                 // scope 2 at src/main.rs:20:20: 20:21
        _7 = _3;                         // scope 2 at src/main.rs:20:20: 20:21
        (_5.0: A) = move _6;             // scope 2 at src/main.rs:20:13: 20:23
        (_5.1: B) = move _7;             // scope 2 at src/main.rs:20:13: 20:23
        StorageDead(_7);                 // scope 2 at src/main.rs:20:22: 20:23
        StorageDead(_6);                 // scope 2 at src/main.rs:20:22: 20:23
        StorageLive(_8);                 // scope 3 at src/main.rs:22:16: 22:18
        _8 = move (_5.0: A);             // scope 3 at src/main.rs:22:16: 22:18
        StorageLive(_9);                 // scope 3 at src/main.rs:22:23: 22:25
        _9 = (_5.1: B);                  // scope 3 at src/main.rs:22:23: 22:25
        _0 = const ();                   // scope 0 at src/main.rs:17:11: 23:2
        StorageDead(_9);                 // scope 3 at src/main.rs:23:1: 23:2
        StorageDead(_8);                 // scope 3 at src/main.rs:23:1: 23:2
        StorageDead(_5);                 // scope 2 at src/main.rs:23:1: 23:2
        StorageDead(_3);                 // scope 1 at src/main.rs:23:1: 23:2
        StorageDead(_1);                 // scope 0 at src/main.rs:23:1: 23:2
        return;                          // scope 0 at src/main.rs:23:2: 23:2
    }
}
```

i didn't expect that either. also there might be further optimizations down the line when the MIR gets compiled to LLVM IR. i guess the only real take away is, that one should only derive `Copy` if the type is actually small enough to be copied^^

Reply:
>one should only derive `Copy` if the type is actually small enough to be copied^^

nice take away. thanks for looking into this, in detail. Deserves a post of its own :+1: